pubsublite/quickstart_publisher/main.go

```diff
@@ -48,33 +54,45 @@ func main() {
 	defer publisher.Stop()
 
 	// Publish messages. Messages are automatically batched.
-	var results []*pubsub.PublishResult
+	var published []*publishedMessage
 	for i := 0; i < *messageCount; i++ {
-		r := publisher.Publish(ctx, &pubsub.Message{
+		msg := &pubsub.Message{
 			Data: []byte(fmt.Sprintf("message-%d", i)),
-		})
-		results = append(results, r)
+		}
+		result := publisher.Publish(ctx, msg)
+		published = append(published, &publishedMessage{msg, result})
 	}
 
-	// Print publish results.
-	for _, r := range results {
+	// Print publish results and collect any messages that need to be republished.
+	var toRepublish []*pubsub.Message
+	for _, pm := range published {
 		// Get blocks until the result is ready.
-		id, err := r.Get(ctx)
+		id, err := pm.result.Get(ctx)
```

Comment:
I think if I were writing this, I would use https://pkg.go.dev/golang.org/x/sync/errgroup. But, that's probably too complex for a sample, and I don't think we've done it before.

In the same vein, an alternative would be to do `publisher.Publish(...).Get()` in a goroutine and put results/errors into a channel/map. Then you wouldn't need the `publishedMessage` type.

I'm not sure that's better. But, I wanted to through a couple alternatives out there to see what you thought. I'm OK to move forward as-is.

Reply:
I took a look at some of pubsub's samples, and there are some that use `sync.WaitGroup` (e.g. [publish_ordering.go](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/pubsub/topics/publish_ordering.go)), while others loop a slice of PublishResults (e.g. [publish_settings.go](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/pubsub/topics/publish_settings.go)).

I prefer `errgroup.Group` over `sync.WaitGroup`, and have amended this quickstart so that we have at least one realistic example of error handling.